text/3721-homogeneous-try-blocks.md

```diff
@@ -0,0 +1,759 @@
+
+## Supporting generics
+
+Essentially the same as the previous section, but this doesn't work on nightly either:
+
+<!-- https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fead74266baf2900be2bf69a61fbe618 -->
+
+```rust
+let x = try { slice.get(i)? + slice.get(j)? };
+dbg!(x);
+```
+
+because `dbg!` accepts any `Debug`gable type and thus here it also doesn't know what type you want
+
+```
+error[E0282]: type annotations needed
+ --> src/lib.rs:4:9
+  |
+4 |     let x = try { slice.get(i)? + slice.get(j)? };
+  |         ^
+5 |     dbg!(x);
+  |          - type must be known at this point
+  |
+help: consider giving `x` an explicit type
+  |
+4 |     let x: /* Type */ = try { slice.get(i)? + slice.get(j)? };
+  |          ++++++++++++
+```
+
+Homogeneous `try` fixes this as well.
+
+## The simple case deserves the simple syntax
+
+We could add a new `try homogeneous { ... }` block with this behaviour, and leave `try { ... }` as heterogeneous.
+
+That feels backwards, because heterogeneous try blocks are the ones that most commonly need a *type* annotation of some sort.
+
+If there's `?`s on multiple `Result`s with incompatible error types, we need to tell it *somehow* which type to use.
+Maybe we want an `anyhow::Result<_>`, maybe we want our own `Result<_, crate::CustomError>`, whatever.
+
+Thus if they commonly need a type annotation anyway, we can consider in the future (see below for more)
+an annotated version of `try` blocks that allow heterogeneity, while leaving the short thing for the simple case.
+
+## Manual error conversion is always possible
+
+Even inside a homogeneous `try` block, you could always *manually* add a call to convert an error.
+
+For example, you could do something like
+```rust
+try {
+    foo()?;
+    bar().map_err(Into::into)?;
+    qux()?;
+}
+```
+
+if you need to convert the error type from `bar` to the one used by `foo` and `qux`.
+
+We could always add a specific method to express that intent, though this RFC does not propose one.
+Spelling it as `.map_err(into)` might be pretty good already, which would be possible with [RFC#3591].
+
+[RFC#3591]: https://github.com/rust-lang/rfcs/pull/3591
+
+## Other merging approaches
+
+There's a variety of other things we could do if the `?`s don't all match.
+
+- Maybe we try to convert everything to the first one
+- Maybe we try to convert everything to the last one
+- Maybe we fold them through some type function that attempts to merge residuals
+
+But these are all much less local.
+
+A nice property of the homogeneous `try` block is that you don't have to think about all this stuff.
+When you see `try {`, you know that they're all the same.  You can thus reorder them without worrying.
+So long as you know what family one of them is from, you know the rest are the same.
+
+## This case really is common
+
+The rust compiler uses `try` blocks in a bunch of places already.  Last I checked, they were *all* homogeneous.
```

Comment:
I don't know if the compiler is the ideal place to look for evaluating if this case is common or not, as it has very structured code that focuses on a single domain. The examples show essentially only situations where the try block is applied only on options or only on calls that return `io::Result`. I think that the compiler mostly uses similar types for errors, rather than wrappers like `Box<dyn Error>` (AFAIK). But in my experience, Rust code in the wild tends to be much more messy, especially in applications.

In particular, I think there is one use-case that should IMO by supported out-of-the-box by `try` blocks, but I'm not sure how well it would work with homogeneous try blocks, any that is integration with `anyhow::Error` (as its usage is *super common* in applications, at least in my experience). For example, a sequence of fallible operations that I encounter all the time in Rust applications (so in very *applied* messy code that deals with a lot of very different fallible things at once) is trying to fetch something from an API (`io::error`), then deserializing it (`serde::Error`) and then e.g. checking some domain invariant, or storing the thing inside a DB (which both would produce a different kind of an error). These cases are then all unified into `anyhow::Error`. This works perfectly with `?` in functions today, but would be quite annoying to do with homogeneous try blocks. Arguably, just spamming `?` might not be ideal, but that can be improved with `.context/.with_context`. It would be quite verbose and boilerplate-y if we had to write `foo()[.context("...")].map_err(Into::into)?` after every such operation. `map_err` is great for carefully written library-like code, not so much for high-level application-like code that just wants to get stuff done with the least amount of cruft.

That being said, annotating the try blocks such as `try $ anyhow::Error` seems like a reasonable solution to fix this.

Reply:
Ah, good point, I didn't realize that context returns anyhow::Error, but it makes sense of course.

That being said, even though using context is the "proper thing to do" in many (but not all) situations, I do see (and write) code that does indeed just spam ?.

In fact, for me, the try operator is a way to avoid using context everywhere! Usually when I want to use try, it's because I essentially want to group a few errors together, and treat them as one opaqur error (to which I then attach the context), but I don't care about the specific errors within the try block so much. So something like `try { a?; b?; }.context(...)`. If the try block could figure out from that context call that I want anyhow::Error, that would also be nice (I saw the discussion about something like this in the RFC).

Yeah, I saw the future possibility with the return type annotation, that's why I mentioned it :)

Reply:
For folks using `anyhow`, needing to annotate the return type of a block isn't a huge deal because the library has an alias, [`anyhow::Ok`](<https://docs.rs/anyhow/latest/anyhow/fn.Ok.html>). In the projects I've worked in using anyhow, it's a huge lifesaver for closures that return `anyhow::Result<T>`.

I know that there's probably consensus on implicit `Ok`-wrapping in `try` blocks, but explicitly spelling it out and having an `Ok` alias like `anyhow` would be the sweet path for me and the projects I work on. Basically:

```rust
try {
    let saves = foo()?.bar()?;
    self.cache_savegames(saves)?;
    anyhow::Ok(())
}.context("failed to load savegame list")?;
```

We also have a lot of code that uses `anyhow` with heterogeneous error types. Oftentimes our `.context` call happens _outside_ the big block of `?` spam, like when a function has a bunch of small bits of I/O.